### PR TITLE
Cryptodev linux

### PR DESCRIPTION
**Pull request by nmav**
This pull request corrects issues with the current engine for /dev/crypto.

Also in:
http://rt.openssl.org/Ticket/Display.html?id=2770&user=guest&pass=guest

### CODE REVIEW

**Comment by sreimers** [comment]
+1

**Comment by flowher** [comment]
+1

**Comment by richsalz** [comment]
is this still an issue?

**Comment by nmav** [comment]
To be honest I no longer work with cryptodev. Maybe someone else would take on the patches if he's interested on the issue.

**Comment by mattcaswell** [comment]
@richsalz. The OP has moved on, and no one else has stepped up to own these patches. They would need some significant work now to bring them up to date with all the latest opacity changes etc. I think we should close this. Do you agree?

**Comment by richsalz** [comment]
Isn’t Ben interested in this?  Not saying don't close, but perhaps ask him.

**Comment by mattcaswell** [comment]
Ping @benlaurie. Your opinion?

**Comment by levitte** [comment]
I'm interested in this and can take the lead.  However, this should be made a post 1.1.0 feature request.

**Comment by moonman** [comment]
Because of refactoring the patches don't apply with newer version of openssl. I adjusted Nmav's patch and it is available here: https://github.com/archlinuxarm/PKGBUILDs/blob/master/core/openssl-cryptodev/fix_sig-gen_and_allow_copy_evp_contexts.patch

**Comment by Voker57** [comment]
I also adjusted the patch and improved it by enabling more algorithms and styling the code: https://github.com/Voker57/openssl/tree/cryptodev-patches

**Comment by richsalz** [comment]
If anyone is interested in submitting a new PR for this, that would be great.  Please see the CONTRIBUTING file for CLA etc.

File: crypto/engine/eng_cryptodev.c [hunk 1]
```diff
@@ -294,12 +301,12 @@ get_cryptodev_digests(const int **cnids)
...
 	}
 	memset(&sess, 0, sizeof(sess));
-	sess.mackey = (caddr_t)"123456789abcdefghijklmno";
+	sess.mackey = fake_key;
 	for (i = 0; digests[i].id && count < CRYPTO_ALGORITHM_MAX; i++) {
 		if (digests[i].nid == NID_undef)
 			continue;
 		sess.mac = digests[i].id;
-		sess.mackeylen = digests[i].keylen;
+		sess.mackeylen = 8;
```
Comment (levitte) [inline crypto/engine/eng_cryptodev.c]: Hmmm?  Shouldn't this be `digests[i].digestlen`?

File: crypto/engine/eng_cryptodev.c [hunk 2]
```diff
@@ -923,34 +928,117 @@ static int cryptodev_digest_copy(EVP_MD_CTX *to,const EVP_MD_CTX *from)
...
+	EVP_MD_FLAG_DIGALGID_ABSENT|
+#endif
 	EVP_MD_FLAG_ONESHOT,
 	cryptodev_digest_init,
 	cryptodev_digest_update,
 	cryptodev_digest_final,
 	cryptodev_digest_copy,
 	cryptodev_digest_cleanup,
-	EVP_PKEY_NULL_method,
+	EVP_PKEY_RSA_method,
```
Comment (levitte) [inline crypto/engine/eng_cryptodev.c]: This is not easily transferable to master at this point.  EVP_MD doesn't hold those fields any more, and there are EVP_PKEY methods instead

**Comment by levitte** [comment]
Closing this, as it became part of #1784 (through #1792)

**Pull request closed without merging**